File: app/src/test/java/com/starsep/sokoban/release/gamelogic/TileTest.java
package com.starsep.sokoban.release.gamelogic;

import android.util.Log;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Log.class)
public class TileTest {
	@Test
	public void crateIsCrate() {
		assertTrue(Tile.INSTANCE.isCrate(Tile.INSTANCE.getCrate()));
	}

	@Test
	public void crateIsNotWalkable() {
		assertFalse(Tile.INSTANCE.isWalkable(Tile.INSTANCE.getCrate()));
	}

	@Test
	public void groundAndGrassIsWalkable() {
		assertTrue(Tile.INSTANCE.isWalkable(Tile.INSTANCE.getGround()));
		assertTrue(Tile.INSTANCE.isWalkable(Tile.INSTANCE.getGrass()));
	}

	@Test
	public void maskToMaskToCharIsIdentity() {
		char[] tiles = {
                Tile.INSTANCE.getCrate(), Tile.INSTANCE.getGrass(), Tile.INSTANCE.getGround(), Tile.INSTANCE.getEndpoint(), Tile.INSTANCE.getCrateOnEndpoint(), Tile.INSTANCE.getWall()};
		for (char tile : tiles) {
			assertEquals(Tile.INSTANCE.maskToChar(Tile.INSTANCE.mask(tile)), tile);
		}
	}

	@Test
	public void defaultMaskWalkable() {
		PowerMockito.mockStatic(Log.class);
		assertEquals(Tile.INSTANCE.mask('~'), Tile.INSTANCE.getWALKABLE_MASK());
		PowerMockito.verifyStatic();
	}

	@Test
	public void defaultMaskToCharGrass() {
		PowerMockito.mockStatic(Log.class);
		assertEquals(Tile.INSTANCE.maskToChar(1337), Tile.INSTANCE.getGrass());
		PowerMockito.verifyStatic();
	}
}

File: app/src/test/java/com/starsep/sokoban/release/gamelogic/TileTest.kt
package com.starsep.sokoban.release.gamelogic

import com.starsep.sokoban.release.gamelogic.Tile.WALKABLE_MASK
import com.starsep.sokoban.release.gamelogic.Tile.crate
import com.starsep.sokoban.release.gamelogic.Tile.crateOnEndpoint
import com.starsep.sokoban.release.gamelogic.Tile.endpoint
import com.starsep.sokoban.release.gamelogic.Tile.grass
import com.starsep.sokoban.release.gamelogic.Tile.ground
import com.starsep.sokoban.release.gamelogic.Tile.isCrate
import com.starsep.sokoban.release.gamelogic.Tile.isWalkable
import com.starsep.sokoban.release.gamelogic.Tile.mask
import com.starsep.sokoban.release.gamelogic.Tile.maskToChar
import com.starsep.sokoban.release.gamelogic.Tile.wall
import org.junit.Assert
import org.junit.Test

class TileTest {
    @Test
    fun crateIsCrate() {
        Assert.assertTrue(isCrate(crate))
    }

    @Test
    fun crateIsNotWalkable() {
        Assert.assertFalse(isWalkable(crate))
    }

    @Test
    fun groundAndGrassIsWalkable() {
        Assert.assertTrue(isWalkable(ground))
        Assert.assertTrue(isWalkable(grass))
    }

    @Test
    fun maskToMaskToCharIsIdentity() {
        val tiles = charArrayOf(
            crate,
            grass,
            ground,
            endpoint,
            crateOnEndpoint,
            wall
        )
        for (tile in tiles) {
            Assert.assertEquals(
                maskToChar(
                    mask(
                        tile
                    )
                ),
                tile
            )
        }
    }

    @Test
    fun defaultMaskWalkable() {
        Assert.assertEquals(
            mask('~'),
            WALKABLE_MASK
        )
    }

    @Test
    fun defaultMaskToCharGrass() {
        Assert.assertEquals(
            maskToChar(1337),
            grass
        )
    }
}
